### ISSUE TITLE
Suggestion: Setting to load environment from file, for safe secret handling

### ISSUE THREAD

**Issue by trbabb** [Supervisor/supervisor#1304]
Currently, the handling of application secrets under `supervisord` is inflexible and potentially unsafe. If passed as environment variables (a standard practice), the secrets must live in the supervisord configuration file itself, which is not particularly secure (it might need to be checked into source control, e.g.; or be readable to non-root users).

It would be useful if there were a setting parallel to the `environment` key which could take environments from the filesystem, e.g. `environment_file=/path/to/resource`, where the target file contains `KEY=value` pairs, one per line.

This way, the environment file can be owned by `root`, and have `-r--------` permissions; then `supervisord` could read the file and build its environment as root *before* reducing permissions for the daemonized application. This way, the secrets don't have to be saved in configurations (thus inheriting their permissions), or be readable to the child process, or possibly be checked into source repos.

There is precedent for this in (for example) `uwsgi`, which has the `--envdir` option, which essentially functions as above; except keys are filenames in the provided directory, and the values are the file contents— this could be another way to expose this functionality which might be more friendly to multi-line variable values.

**Comment by trbabb** [comment]
For further background, [here](https://stackoverflow.com/questions/45903060/how-to-pass-shell-environment-variable-to-supervisor-program) is an example of another user encountering this problem, where the suggested workaround is to write a wrapper shell script that reads the file and then invokes the target program.

However, this has the problem that the child process is already running with reduced permissions, so the environment file can't be properly protected. To protect the file, the daemon service would instead have to run as `root` (dangerous), and the wrapper script would need to duplicate `supervisord`'s privilege de-escalation (also dangerous, and too much work, and likely to be done incorrectly).

**Comment by mnaberez** [comment]
Duplicate of #242

**Comment by trbabb** [comment]
Thanks for considering; it seems like this a much-wanted feature.

Any comment on the concerns above re: security of a wrapper script?

**Comment by mnaberez** [comment]
Before starting `supervisord` in your init script, put the variables into the environment in a way you feel is secure.  `supervisord` inherits the environment of the shell that starts it, and subprocesses running under `supervisord` inherit it as well.

**Comment by trbabb** [comment]
Wouldn't that expose the secrets to all supervisord processes, not just the app that needs the secret?

**Comment by mnaberez** [comment]
From [supervisord.org](http://supervisord.org/):

> It shares some of the same goals of programs like launchd, daemontools, and runit. Unlike some of these programs, it is not meant to be run as a substitute for init as “process id 1”. Instead it is meant to be used to control processes related to a project or a customer, and is meant to start like any other program at boot time.

It sounds like you want a single `supervisord` to be more like an init system and run several applications, but it was originally intended to run one.

**Comment by trbabb** [comment]
It looked to me like `supervisord` explicitly supports multiple program configs at once, but maybe I misread. Is that wrong?

**Comment by mnaberez** [comment]
We use the word `program` in the configuration file instead of `process` because each `[program]` section can spawn multiple processes using the `numprocs` option.

**Comment by trbabb** [comment]
Right. So if I define multiple `[program]`s, then they will *all* see each others' secrets, even if any given secret is only needed by one?

**Comment by mnaberez** [comment]
You could run multiple `supervisord` instances, where each controls the processes for one isolated application.  The secrets for each app would be defined in the environment before each `supervisord` instance was started.  This was the original use of `supervisord`.  It implies that the secrets can be shared among those processes of the application.  If none of these options work, it may be that `supervisord` is not the right tool for your use case.
